src/hotspot/share/gc/shenandoah/shenandoahLock.hpp

```diff
@@ -37,21 +37,26 @@ class ShenandoahLock  {
   shenandoah_padding(0);
   volatile LockState _state;
   shenandoah_padding(1);
-  volatile Thread* _owner;
+  Thread* volatile _owner;
   shenandoah_padding(2);
 
-  template<typename BlockOp>
+  template<bool ALLOW_BLOCK>
   void contended_lock_internal(JavaThread* java_thread);
-
 public:
   ShenandoahLock() : _state(unlocked), _owner(nullptr) {};
 
   void lock(bool allow_block_for_safepoint) {
     assert(Atomic::load(&_owner) != Thread::current(), "reentrant locking attempt, would deadlock");
 
-    // Try to lock fast, or dive into contended lock handling.
-    if (Atomic::cmpxchg(&_state, unlocked, locked) != unlocked) {
+    if (allow_block_for_safepoint && SafepointSynchronize::is_synchronizing()) {
+      // Java thread, and there is a pending safepoint. Dive into contended locking
+      // immediately without trying anything else, and block.
       contended_lock(allow_block_for_safepoint);
+    } else {
+      // Try to lock fast, or dive into contended lock handling.
+      if (Atomic::cmpxchg(&_state, unlocked, locked) != unlocked) {
+        contended_lock(allow_block_for_safepoint);
+      }
```

Comment:
I am thinking we can increase the chances for inline if we collapse both branches like:

```
  // Dive into the contended locking handling if there is a safepoint pending and we can
  // block the thread, or the fast-path locking have failed.
  if ((allow_block_for_safepoint && SafepointSynchronize:is_synchronizing()) ||
      (Atomic::cmpxchg(&_state, unlocked, locked) != unlocked)) {
    contended_lock(allow_block_for_safepoint);
  }
```